pkg/winsvc/winsvc.go

```diff
@@ -0,0 +1,175 @@
+//go:build windows
+
+package winsvc
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"golang.org/x/sys/windows/svc"
+	"golang.org/x/sys/windows/svc/mgr"
+)
+
+type Mgr interface {
+	CreateService(string, string, mgr.Config, ...string) (Service, error)
+	ListServices() ([]string, error)
+	OpenService(string) (Service, error)
+}
+
+type fakeServiceList struct {
+	m    *sync.Mutex
+	svcs map[string]Service
+}
+
+// write overwrites the given service to the svcs map
+func (l *fakeServiceList) write(name string, svc Service) {
+	l.m.Lock()
+	defer l.m.Unlock()
+	l.svcs[name] = svc
+}
+
+// read returns the entry with the given name, and a bool indicating if it exists or not
+func (l *fakeServiceList) read(name string) (Service, bool) {
+	l.m.Lock()
+	defer l.m.Unlock()
+	service, exists := l.svcs[name]
+	return service, exists
+}
+
+// listServiceNames returns a slice of all service names
+func (l *fakeServiceList) listServiceNames() []string {
+	l.m.Lock()
+	defer l.m.Unlock()
+	var names []string
+	for svcName := range l.svcs {
+		names = append(names, svcName)
+	}
+	return names
+}
+
+// remove deletes the entry with the given name, throwing an error if it doesn't exist
+func (l *fakeServiceList) remove(name string) error {
+	l.m.Lock()
+	defer l.m.Unlock()
+	_, exists := l.svcs[name]
```

Comment:
Why not using `read()`?

Reply:
read cannot be called as that would cause a deadlock due to both remove and read locking.